peer.go

```diff
@@ -0,0 +1,172 @@
+}
+
+var calibnetPeers = []string{
+	"/dns4/bootstrap-0.calibration.fildev.network/tcp/1347/p2p/12D3KooWJkikQQkxS58spo76BYzFt4fotaT5NpV2zngvrqm4u5ow",
+	"/dns4/bootstrap-1.calibration.fildev.network/tcp/1347/p2p/12D3KooWLce5FDHR4EX4CrYavphA5xS3uDsX6aoowXh5tzDUxJav",
+	"/dns4/bootstrap-2.calibration.fildev.network/tcp/1347/p2p/12D3KooWA9hFfQG9GjP6bHeuQQbMD3FDtZLdW1NayxKXUT26PQZu",
+	"/dns4/bootstrap-3.calibration.fildev.network/tcp/1347/p2p/12D3KooWMHDi3LVTFG8Szqogt7RkNXvonbQYqSazxBx41A5aeuVz",
+}
+
+type filecoinPeer struct {
+	ctx            context.Context
+	dht            *dht.IpfsDHT
+	host           host.Host
+	bootstrapPeers []peer.AddrInfo
+	foundMiners    []peer.AddrInfo
+}
+
+func filecoinBootstrapPeers(peerList []string) []peer.AddrInfo {
+	if peerList == nil {
+		peerList = mainnetPeers
+	}
+
+	maddrs := make([]multiaddr.Multiaddr, len(peerList))
+
+	for i, bp := range peerList {
+		var err error
+		maddrs[i], err = multiaddr.NewMultiaddr(bp)
+		if err != nil {
+			log.Fatal(err)
+			continue
+		}
+	}
+
+	peers, _ := peer.AddrInfosFromP2pAddrs(maddrs...)
+
+	return peers
+}
+
+func setupFilecoinPeer(
+	ctx context.Context,
+	network networkName,
+) (*filecoinPeer, error) {
+	if network == "" {
+		network = mainnet
+	}
+
+	var pp protocol.ID
+	var bootstrapPeers []peer.AddrInfo
+
+	switch network {
+	case mainnet:
+		pp = "/fil/kad/testnetnet"
+		bootstrapPeers = filecoinBootstrapPeers(mainnetPeers)
+	case calibnet:
+		pp = "/fil/kad/calibnet"
+		bootstrapPeers = filecoinBootstrapPeers(calibnetPeers)
+	}
+
+	var ddht *dht.IpfsDHT
+	priv, _, err := crypto.GenerateKeyPair(crypto.Ed25519, -1)
+
+	if err != nil {
+		return &filecoinPeer{}, err
+	}
+
+	libp2pOpts := []libp2p.Option{
+		libp2p.Identity(priv),
+		libp2p.Security(libp2ptls.ID, libp2ptls.New),
+		libp2p.Security(noise.ID, noise.New),
+		libp2p.DefaultTransports,
+		libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
+			ddht, err = dht.New(
+				ctx,
+				h,
+				dht.Datastore(dsync.MutexWrap(ds.NewMapDatastore())),
+				dht.NamespacedValidator("pk", record.PublicKeyValidator{}),
+				dht.NamespacedValidator("ipns", ipns.Validator{KeyBook: h.Peerstore()}),
+				dht.Concurrency(50),
+				dht.Mode(dht.ModeClient),
+				dht.ProtocolPrefix(pp),
+			)
+
+			return ddht, err
+		}),
+	}
+
+	h, err := libp2p.New(libp2pOpts...)
+
+	if err != nil {
+		return &filecoinPeer{}, err
+	}
+
+	return &filecoinPeer{ctx: ctx, host: h, dht: ddht, bootstrapPeers: bootstrapPeers}, nil
+}
+
+func (f *filecoinPeer) bootstrap() {
+	connected := make(chan struct{})
+
+	var wg sync.WaitGroup
+	for _, pinfo := range f.bootstrapPeers {
+		p := pinfo
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			err := f.host.Connect(f.ctx, p)
+			if err != nil {
+				log.Warn(err)
+				return
+			}
+			log.Infow("Connected", "ID", p.ID)
+			connected <- struct{}{}
+		}()
+	}
+
+	go func() {
+		wg.Wait()
+		close(connected)
```

Comment:
we might want to call dht.Bootstrap() (though I'm suspecting it happens automatically in the background).